src/EFCore/Metadata/Internal/InternalModelBuilder.cs

```diff
@@ -211,6 +212,191 @@ public InternalModelBuilder([NotNull] Model metadata)
+            if (rightForeignKey == null)
+            {
+                // Removing the association entity type will also remove
+                // the leftForeignKey created above.
+                HasNoEntityType(associationEntityTypeBuilder.Metadata, configurationSource);
+
+                if (configurationSource != ConfigurationSource.Explicit)
+                {
+                    return null;
+                }
+
+                throw new InvalidOperationException(
+                    CoreStrings.UnableToCreateSkipNavigationForeignKeyOnAssociationEntityType(
+                        rightEntityType.Name,
+                        rightSkipNavigation.Name,
+                        associationEntityTypeName));
+            }
+
+            leftSkipNavigation.SetForeignKey(leftForeignKey, configurationSource);
+            rightSkipNavigation.SetForeignKey(rightForeignKey, configurationSource);
+            leftSkipNavigation.Builder.HasInverse(rightSkipNavigation, configurationSource);
+
+            // Creating the primary key below also negates the need for an index on
+            // the properties of leftForeignKey - that index is automatically removed.
+            associationEntityTypeBuilder.PrimaryKey(
+                leftForeignKey.Properties.Concat(rightForeignKey.Properties).ToList(),
+                configurationSource);
+
+            return associationEntityTypeBuilder;
+        }
+
+        private static ForeignKey CreateSkipNavigationForeignKey(
+            SkipNavigation skipNavigation,
+            InternalEntityTypeBuilder associationEntityTypeBuilder,
+            ConfigurationSource configurationSource)
+        {
+            var principalEntityType = skipNavigation.DeclaringEntityType;
+            var principalKey = principalEntityType.FindPrimaryKey();
+            if (principalKey == null)
+            {
+                return null;
+            }
+
+            var dependentEndForeignKeyPropertyNames = new List<string>();
+            var otherIdentifiers = associationEntityTypeBuilder.Metadata
+                .GetDeclaredProperties().ToDictionary(p => p.Name, p => 0);
+            foreach (var property in principalKey.Properties)
+            {
+                var propertyName = Uniquifier.Uniquify(
+                    string.Format(
+                        AssociationPropertyNameTemplate,
+                        principalEntityType.ShortName(),
+                        property.Name),
+                    otherIdentifiers,
+                    int.MaxValue);
+                dependentEndForeignKeyPropertyNames.Add(propertyName);
+                otherIdentifiers.Add(propertyName, 0);
+            }
+
+            return associationEntityTypeBuilder
+                .HasRelationship(
+                    principalEntityType.Name,
+                    dependentEndForeignKeyPropertyNames,
+                    principalKey,
+                    configurationSource)
```

Comment:
Let the builder figure out the details:
```suggestion
                .HasRelationship(principalEntityType, configurationSource)
```

Reply:
Also configure the resulting FK as non-unique

Reply:
I'd rather call the specific one - I know what properties I want to use, I know what I want the principal key to be - I'd rather call that and if it errors I know what the problem is rather than having the other API take it's best guess and possibly fail, or even worse succeed but set the wrong properties/key.

Is the purpose of setting FK to non-unique to update its `ConfigurationSource`? Otherwise I think it's would be non-unique by default.

Reply:
You are just creating the properties to use and we already have a pattern for how they are created, we don't need to duplicate the code. There are no "wrong" properties/key here
The other overload will use the primary key if it exists and create a temporary one if it doesn't, it will only return `null` if the principal entity type is configured as keyless.

> Is the purpose of setting FK to non-unique to update its ConfigurationSource? Otherwise I think it's would be non-unique by default.

Yes

Reply:
I'll do the non-unique thing.

The primary key comment is moot - I check that the PK exists (I thought it would error if it didn't exist but it doesn't matter). The real difference is in the naming of the FK properties - which I'm happy to change if the team decides they don't like the naming approach I suggested.